File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n=nums.size();

        vector<int> ans(n);

        ans[0]=1;
        for(int i=1;i<n;i++){
            ans[i]=nums[i-1]*ans[i-1];
        }

       int product=nums[n-1];
        for(int i=n-2;i>=0;i--){ 
            ans[i]=ans[i]*product;
            product*=nums[i];
        }

        return ans;

    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
       int[] left = new int[nums.length];

        left[0] = 1;
        for(int i = 1; i < nums.length; i++){
            left[i] = left[i -1] * nums[i-1] ;
        }
        
        int[] result = new int[nums.length]; 
        int multiple = nums[nums.length -1];
        result[nums.length -1] = left[left.length -1];
        for(int i = nums.length-2 ; i >=0 ; i--){
            result[i] = multiple * left[i];
            multiple = multiple * nums[i];
        }

        return result;
    }
}